packages/mui-joy/src/Button/Button.tsx

```diff
@@ -1,21 +1,250 @@
+      position: 'relative',
+      ...theme.typography.button,
+      [`&.${buttonClasses.disabled}`]: {
+        pointerEvents: 'none',
+        cursor: 'default',
+      },
+      [`&.${buttonClasses.focusVisible}`]: {
+        outline: '4px solid',
+        outlineColor: colorPalette[300],
+      },
+      ...(ownerState.fullWidth && {
+        width: '100%',
+      }),
+    },
+    ownerState.size === 'small' && {
+      minHeight: '40px',
+    },
+    ownerState.size === 'large' && {
+      minHeight: '56px',
+    },
+    ownerState.variant === 'text' && {
+      color: colorPalette[600],
+      [`&.${buttonClasses.focusVisible}`]: {
+        outlineColor: neutral[200],
+      },
+      '&:hover': {
+        backgroundColor: neutral[100],
+      },
+      '&:active': {
+        backgroundColor: neutral[200],
+      },
+      [`&.${buttonClasses.disabled}`]: {
+        color: neutral[300],
+      },
+    },
+    ownerState.variant === 'contained' && {
+      backgroundColor: colorPalette[600],
+      color: '#fff',
+      '&:hover': {
+        backgroundColor: colorPalette[700],
+      },
+      '&:active': {
+        backgroundColor: colorPalette[500],
+      },
+      [`&.${buttonClasses.disabled}`]: {
+        backgroundColor: colorPalette[300],
+      },
+    },
+    ownerState.variant === 'outlined' && {
+      color: colorPalette[600],
+      border: '1px solid',
+      borderColor: neutral[300],
+      [`&.${buttonClasses.focusVisible}`]: {
+        outlineColor: neutral[200],
+      },
+      '&:hover': {
+        backgroundColor: neutral[100],
+      },
+      '&:active': {
+        backgroundColor: neutral[200],
+      },
+      [`&.${buttonClasses.disabled}`]: {
+        borderColor: neutral[200],
+        color: neutral[300],
+      },
+    },
+  ];
+});
+
+const Button = React.forwardRef(function Button(inProps, ref) {
+  const props = useThemeProps({ props: inProps, name: 'JoyButton' });
+
+  const {
+    children,
+    className,
+    action,
+    component = 'button',
+    color = 'brand',
+    variant = 'text',
+    size = 'medium',
+    fullWidth = false,
+    ...other
+  } = props;
+
+  const buttonRef = React.useRef<HTMLElement | null>(null);
+  const handleRef = useForkRef(buttonRef, ref);
+
+  const ComponentProp = component;
+
+  const { focusVisible, setFocusVisible, getRootProps } = useButton({
+    ...props,
+    component: ComponentProp,
+    ref: handleRef,
+  });
+
+  React.useImperativeHandle(
+    action,
+    () => ({
+      focusVisible: () => {
+        setFocusVisible(true);
+        buttonRef.current?.focus();
+      },
+    }),
+    [setFocusVisible],
+  );
+
+  const ownerState = {
+    ...props,
+    component,
+    color,
+    fullWidth,
+    variant,
+    size,
+    focusVisible,
+  };
+
+  const classes = useUtilityClasses(ownerState);
+
+  return (
+    <ButtonRoot
+      as={ComponentProp}
+      className={clsx(classes.root, className)}
+      ownerState={ownerState}
```

Comment:
@michaldudak we discussed that we may want to rename these states to `buttonState` so that when we have conflict when used in other components. Just mentioning it here so that we consider it in this early stage.

Reply:
I'm going to work on changing the ownerState to [component]State even this week, so @siriwatknp we'll have to update this PR as well (unless it lands on master first)

Reply:
It could be tricky as `ownerState` is kind of considered in some of the system utils too, so watch out for that too

Reply:
In the end, I ended up using the hook instead of the unstyled component in my base. I think this scales better and we will avoid issues like this. If there is conflict in names we can be explicit when creating the `ownerProps` in the outer component, for example prefixing the props coming from the button with button and have `ownerState.buttonActive`